### PR TITLE
Automated helm deploys (A.K.A. helm addons)

### LINKED ISSUE #2441: helm addons implementation
Write implementation of the [helm addons proposal](https://github.com/kubermatic/kubeone/blob/main/docs/proposals/20221021-helm-addons.md).

### PR DESCRIPTION
**Pull request by kron4eg**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2441 

**What type of PR is this?**
/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
```release-note
Automated helm deploys (A.K.A. helm addons)
```

**Documentation**:
```documentation
TBD
```

### CODE REVIEW

File: pkg/tasks/cni.go [hunk 1]
```diff
@@ -1,47 +0,0 @@
...
-import (
-	"github.com/pkg/errors"
-
-	"k8c.io/kubeone/pkg/fail"
-	"k8c.io/kubeone/pkg/state"
-
-	"sigs.k8s.io/controller-runtime/pkg/client"
-)
-
-func ensureCNI(s *state.State) error {
```
Comment (kron4eg) [inline pkg/tasks/cni.go:28]: this file is some leftover. it's doing nothing useful :shrug:

**Review by xmudrii — COMMENTED** [review]
Some initial comments, will continue reviewing later.

File: pkg/apis/kubeone/v1beta2/types.go [hunk 2]
```diff
@@ -84,6 +87,20 @@ type KubeOneCluster struct {
...
+	Chart       string       `json:"chart"`
+	RepoURL     string       `json:"repoURL"`
+	Version     string       `json:"version,omitempty"`
+	ReleaseName string       `json:"releaseName,omitempty"`
+	Namespace   string       `json:"namespace"`
+	Values      []HelmValues `json:"values,omitempty"`
+}
+
+type HelmValues struct {
+	File   string          `json:"file,omitempty"`
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/v1beta2/types.go]: Renamed to ValuesFile with a doc comment about file locality

File: pkg/apis/kubeone/v1beta2/types.go [hunk 3]
```diff
@@ -84,6 +87,20 @@ type KubeOneCluster struct {
...
 }
 
+type HelmRelease struct {
+	Chart       string       `json:"chart"`
+	RepoURL     string       `json:"repoURL"`
+	Version     string       `json:"version,omitempty"`
+	ReleaseName string       `json:"releaseName,omitempty"`
+	Namespace   string       `json:"namespace"`
+	Values      []HelmValues `json:"values,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/v1beta2/types.go]: done

File: pkg/apis/kubeone/v1beta2/types.go [hunk 4]
```diff
@@ -74,6 +74,9 @@ type KubeOneCluster struct {
 	// Addons are used to deploy additional manifests.
 	Addons *Addons `json:"addons,omitempty"`
 
+	// HelmReleases list of instructions of how to deploy helm charts
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/v1beta2/types.go]: +

File: Makefile [hunk 5]
```diff
@@ -12,6 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+SHELL := bash
+.ONESHELL:
+.DELETE_ON_ERROR:
+.SHELLFLAGS := -eu -o pipefail -c
+MAKEFLAGS += --warn-undefined-variables
+MAKEFLAGS += --no-builtin-rules
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply Makefile:20]: There are literally no comments in this makefile, why is this something special?

[the remark this answers is not in the document]
Comment (xmudrii) [reply Makefile:20]: > There are literally no comments in this makefile

This is true, but it's something we can and should change. :wink: 

> why is this something special?

This is not something common (at least from my experience), so leaving a short explanation can be very useful to folks reading the Makefile. Some questions that come to my mind:

- Why are we enforcing `bash`?
- Why do we need `.ONESHELL` and `.DELETE_ON_ERROR` rules?
- Why do we need `--no-builtin-rules`? What's exactly doing that one?

File: pkg/apis/kubeone/types.go [hunk 6]
```diff
@@ -87,6 +91,39 @@ type KubeOneCluster struct {
 	LoggingConfig LoggingConfig `json:"loggingConfig,omitempty"`
 }
 
+type HelmRelease struct {
+	// Chart is [CHART] part of the `helm upgrade [RELEASE] [CHART]` command.
+	Chart string `json:"chart"`
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/types.go:96]: The helm upgrade usage looks like this:
```
Usage:
  helm upgrade [RELEASE] [CHART] [flags]
```

There is no [REPO], as "[REPO]" is implied in [CHART].

File: pkg/apis/kubeone/types.go [hunk 7]
```diff
@@ -87,6 +91,39 @@ type KubeOneCluster struct {
...
+
+	// RepoURL chart repository URL where to locate the requested chart.
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/types.go]: :heavy_check_mark:

File: pkg/apis/kubeone/validation/validation.go [hunk 8]
```diff
@@ -645,6 +648,56 @@ func ValidateAddons(o *kubeoneapi.Addons, fldPath *field.Path) field.ErrorList {
 	return allErrs
 }
 
+func ValidateHelmReleases(helmReleases []kubeoneapi.HelmRelease, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for _, hr := range helmReleases {
+		if hr.Chart == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("chart"), hr.Chart, "is a required field"))
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/validation/validation.go]: :heavy_check_mark:

File: pkg/apis/kubeone/validation/validation.go [hunk 9]
```diff
@@ -645,6 +648,56 @@ func ValidateAddons(o *kubeoneapi.Addons, fldPath *field.Path) field.ErrorList {
...
+		}
+
+		if hr.RepoURL == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("repoURL"), hr.RepoURL, "is a required field"))
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/validation/validation.go]: :heavy_check_mark:

File: pkg/apis/kubeone/validation/validation.go [hunk 10]
```diff
@@ -645,6 +648,56 @@ func ValidateAddons(o *kubeoneapi.Addons, fldPath *field.Path) field.ErrorList {
...
+		}
+
+		if hr.Namespace == "" {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("namespace"), hr.Namespace, "is a required field"))
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/apis/kubeone/validation/validation.go]: :heavy_check_mark:

File: pkg/localhelm/helm3.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	if len(st.Cluster.HelmReleases) == 0 {
+		return nil
+	}
+
+	konfigBuf, err := kubeconfig.Download(st)
+	if err != nil {
+		return err
+	}
+
+	tmpKubeConf, err := os.CreateTemp("", "konf*")
```
[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/localhelm/helm3.go]: :heavy_check_mark:

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (kron4eg) [reply Makefile]: :heavy_check_mark:

**Pull request merged**